internal/controllers/machinedeployment/machinedeployment_sync_test.go

```diff
@@ -495,6 +499,224 @@ func TestSyncDeploymentStatus(t *testing.T) {
+		expectedMS.Spec.Template.Labels[clusterv1.MachineDeploymentUniqueLabel] = uniqueID
+
+		g := NewWithT(t)
+		actualMS, err := (&Reconciler{}).computeDesiredMachineSet(deployment, existingMS, nil, log)
+		g.Expect(err).To(BeNil())
+		assertMachineSet(g, actualMS, expectedMS)
+	})
+
+	t.Run("should compute the updated MachineSet when old MachineSets exist", func(t *testing.T) {
+		uniqueID := apirand.String(5)
+		existingMS := skeletonMSBasedOnMD.DeepCopy()
+		existingMSUID := types.UID("abc-123-uid")
+		existingMS.UID = existingMSUID
+		existingMS.Name = deployment.Name + "-" + uniqueID
+		existingMS.Labels = map[string]string{
+			clusterv1.MachineDeploymentUniqueLabel: uniqueID,
+			"ms-label-1":                           "ms-value-1",
+		}
+		existingMS.Annotations = nil
+		existingMS.Spec.Template.Labels = map[string]string{
+			clusterv1.MachineDeploymentUniqueLabel: uniqueID,
+			"ms-label-2":                           "ms-value-2",
+		}
+		existingMS.Spec.Template.Annotations = nil
+		existingMS.Spec.Template.Spec.NodeDrainTimeout = duration5s
+		existingMS.Spec.Template.Spec.NodeDeletionTimeout = duration5s
+		existingMS.Spec.Template.Spec.NodeVolumeDetachTimeout = duration5s
+		existingMS.Spec.DeletePolicy = string(clusterv1.NewestMachineSetDeletePolicy)
+		existingMS.Spec.MinReadySeconds = 0
+
+		oldMS := skeletonMSBasedOnMD.DeepCopy()
+		oldMS.Spec.Replicas = pointer.Int32(2)
+
+		expectedMS := skeletonMSBasedOnMD.DeepCopy()
+		expectedMS.UID = existingMSUID
+		expectedMS.Name = deployment.Name + "-" + uniqueID
+		expectedMS.Labels[clusterv1.MachineDeploymentUniqueLabel] = uniqueID
+		expectedMS.Spec.Template.Labels[clusterv1.MachineDeploymentUniqueLabel] = uniqueID
+
+		g := NewWithT(t)
+		actualMS, err := (&Reconciler{}).computeDesiredMachineSet(deployment, existingMS, []*clusterv1.MachineSet{oldMS}, log)
+		g.Expect(err).To(BeNil())
+		assertMachineSet(g, actualMS, expectedMS)
+	})
+}
+
+func assertMachineSet(g *WithT, actualMS *clusterv1.MachineSet, expectedMS *clusterv1.MachineSet) {
+	// check UID
+	if expectedMS.UID != "" {
+		g.Expect(actualMS.UID).Should(Equal(expectedMS.UID))
+	}
+	// Check Name
+	if expectedMS.Name != "" {
+		g.Expect(actualMS.Name).Should(Equal(expectedMS.Name))
+	}
+	// Check Namespace
+	g.Expect(actualMS.Namespace).Should(Equal(expectedMS.Namespace))
+
+	// Check Replicas
+	g.Expect(actualMS.Spec.Replicas).ShouldNot(BeNil())
+	g.Expect(actualMS.Spec.Replicas).Should(HaveValue(Equal(*expectedMS.Spec.Replicas)))
+
+	// Check ClusterName
+	g.Expect(actualMS.Spec.ClusterName).Should(Equal(expectedMS.Spec.ClusterName))
+
+	// Check Labels
+	for k, v := range expectedMS.Labels {
+		g.Expect(actualMS.Labels).Should(HaveKeyWithValue(k, v))
+	}
+	for k, v := range expectedMS.Spec.Template.Labels {
+		g.Expect(actualMS.Spec.Template.Labels).Should(HaveKeyWithValue(k, v))
+	}
+	// If the expected MS is a new MachineSet then make sure that the labels also has the unique identifier key
```

Comment:
Shouldn't the labels exist in all cases? (if yes let's drop the godoc comment + l.696+699